source/mips/decoder.cpp

```diff
@@ -0,0 +1,410 @@
+/**
+ * @file decoder.cpp
+ *
+ * @brief MIPS Decoder
+ *
+ * @author Zac Brannelly (zac.brannelly@gmail.com)
+ *
+ * SPDX-License-Identifier: GNU General Public License v3.0 only
+ */
+
+#include "decoder.h"
+#include <stdexcept>
+
+static FullOp special_op_table[8][8] = {
+    {FULL_OP_SLL, FULL_OP_INVALID, FULL_OP_SRL, FULL_OP_SRA, FULL_OP_SLLV,
+     FULL_OP_INVALID, FULL_OP_SRLV, FULL_OP_SRAV},
+    {FULL_OP_JR, FULL_OP_JALR, FULL_OP_INVALID, FULL_OP_INVALID,
+     FULL_OP_SYSCALL, FULL_OP_INVALID, FULL_OP_INVALID, FULL_OP_BREAK},
+    {FULL_OP_MFHI, FULL_OP_MTHI, FULL_OP_MFLO, FULL_OP_MTLO, FULL_OP_DSLLV,
+     FULL_OP_INVALID, FULL_OP_DSRLV, FULL_OP_DSRAV},
+    {FULL_OP_MULT, FULL_OP_MULTU, FULL_OP_DIV, FULL_OP_DIVU, FULL_OP_DMULT,
+     FULL_OP_DMULTU, FULL_OP_DDIV, FULL_OP_DDIVU},
+    {FULL_OP_ADD, FULL_OP_ADDU, FULL_OP_SUB, FULL_OP_SUBU, FULL_OP_AND,
+     FULL_OP_OR, FULL_OP_XOR, FULL_OP_NOR},
+    {FULL_OP_INVALID, FULL_OP_INVALID, FULL_OP_SLT, FULL_OP_SLTU, FULL_OP_DADD,
+     FULL_OP_DADDU, FULL_OP_DSUB, FULL_OP_DSUBU},
+    {FULL_OP_TGE, FULL_OP_TGEU, FULL_OP_TLT, FULL_OP_TLTU, FULL_OP_TEQ,
+     FULL_OP_INVALID, FULL_OP_TNE, FULL_OP_INVALID},
+    {FULL_OP_DSLL, FULL_OP_INVALID, FULL_OP_DSRL, FULL_OP_DSRA, FULL_OP_DSLL32,
+     FULL_OP_INVALID, FULL_OP_DSRL32, FULL_OP_DSRA32}};
+
+// A.K.A SPECIAL Function
+Instruction mips_decode_register_type(u32 instruction) {
+  Instruction instruction_data;
+  instruction_data.type = TYPE_REGISTER;
+  populate_register_instruction(&instruction_data.reg, instruction);
+
+  u8 group = (instruction >> 3) & 0x7;
```

Comment:
It might be good to turn these magic constants into something named, but let's do that later :P